daml-lf/spec/daml-lf-1.rst

```diff
@@ -1327,6 +1283,120 @@ Then we define *well-formed expressions*. ::
+type ``τ`` is well-formed, in the rules below. ::
+
+                                ┌──────────────────┐
+  Well-formed case alternatives │ Γ ⊢ τ // alt : σ │
+                                └──────────────────┘
+
+      'variant' T (α₁:k₁) … (αₙ:kₙ) ↦ … | Vᵢ : σᵢ | …  ∈  〚Ξ〛Mod
+      σᵢ  ↠  σᵢ'
+      x : σᵢ'[α₁ ↦ τ₁, …, αₙ ↦ τₙ] · Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltVariant
+      Γ  ⊢  Mod:T τ₁ … τₙ  //  Mod:T:V x  →  e : σ
+
+      'enum' T ↦ … | E | …  ∈  〚Ξ〛Mod
+      Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltEnum
+      Γ  ⊢   Mod:T  //  Mod:T:E  →  e : σ
+
+      Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltNil
+      Γ  ⊢  'List' τ  //  'Nil'  →  e : σ
+
+      xₕ ≠ xₜ
+      xₕ : τ · xₜ : 'List' τ · Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltCons
+      Γ  ⊢  'List' τ  //  'Cons' xₕ xₜ  →  e : σ
+
+      Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltNone
+      Γ  ⊢  'Optional' τ  //  'None'  →  e : σ
+
+      x : τ · Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltSome
+      Γ  ⊢  'Optional' τ  //  'Some' x  →  e : σ
+
+      Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltTrue
+      Γ  ⊢  'Bool'  //  'True'  →  e : σ
+
+      Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltFalse
+      Γ  ⊢  'Bool'  //  'False'  →  e : σ
+
+      Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltUnit
+      Γ  ⊢  'Unit'  //  ()  →  e : σ
+
+      Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltDefault
+      Γ  ⊢  τ  //  _  →  e : σ
+
+
+Pattern match exhaustiveness
+............................
+
+Case expressions ``Γ  ⊢  'case' e 'of' alt₁ | … | altₙ : σ`` also require
+their patterns to be exhaustive, which is defined here. ::
+
+                               ┌─────────────────────┐
+  Pattern match exhaustiveness │ τ  ⊲  alt₁, …, altₙ │
+                               └─────────────────────┘
+
+    'variant' T (α₁:k₁) … (αᵣ:kᵣ) ↦ V₁ : σ₁ | … | Vₘ : σₘ  ∈  〚Ξ〛Mod
+    i₁, i₂, …, iₘ  ∈  {1, …, n}
+    altᵢ₁  =  Mod:T:V₁ x₁  →  e₁
+    altᵢ₂  =  Mod:T:V₂ x₂  →  e₂
+           ⋮
+    altᵢₘ  =  Mod:T:Vₘ xₘ  →  eₘ
+    ——————————————————————————————————————————————————————————————— ExhaustVariant
+    Mod:T τ₁ … τᵣ  ⊲  alt₁, …, altₙ
+
+    'enum' T … ↦ E₁ | … | Eₘ  ∈  〚Ξ〛Mod
+    i₁, i₂, …, iₘ  ∈  {1, …, n}
+    altᵢ₁  =  Mod:T:E₁  →  e₁
+    altᵢ₂  =  Mod:T:E₂  →  e₂
+           ⋮
+    altᵢₘ  =  Mod:T:Eₘ  →  eₘ
+    ——————————————————————————————————————————————————————————————— ExhaustEnum
+    Mod:T  ⊲  alt₁, …, altₙ
+
+    i₁, i₂  ∈  {1, …, n}
+    altᵢ₁  =  'Nil'  →  e₁
+    altᵢ₂  =  'Cons' xₕ xₜ  →  e₂
```

Comment:
Does our font have `j` (or `k`) as an index character? Then we could avoid one level of indexing.

Reply:
Yes, there's a bunch of subscripts. I'll use i,j for these 2-case ones.